Fix stack pointer in core dumps

arbrauns opened the pull request:
The register set in the coredump contains the register values from just before the trap (these are pushed by hardware as the exception stack frame/ESF). Thus, the SP register also needs to point to the stack location from *before* the ESF was pushed; this is simply the location of the ESF plus the size of the ESF.

arbrauns committed:
arch: arm: fix stack pointer in core dumps

The register set in the coredump contains the register values from just
before the trap (these are pushed by hardware as the exception stack
frame/ESF). Thus, the SP register also needs to point to the stack location
from *before* the ESF was pushed; this is obtained by incrementing the
address of the ESF by the size of the ESF.

The ESF can be of two types[0]: a Basic stack frame or an Extended stack
frame (including FPU registers). Bit 4 in EXC_RETURN (stored in lr on
exception entry) determines which type is used.

Cortex-A/Cortex-R, don't have hardware exception stacking at all, but also
no support for coredumps, so the value passed to exc_return is irrelevant.

[0]: https://developer.arm.com/documentation/107706/0100/Floating-point-and-MVE-support/Floating-point-context-handling-mechanisms

Signed-off-by: Armin Brauns <armin.brauns@embedded-solutions.at>

ithinuel commented on arch/arm/core/cortex_m/fault.c:
```diff
@@ -1113,7 +1113,7 @@ void z_arm_fault(uint32_t msp, uint32_t psp, uint32_t exc_return,
 		"ESF could not be retrieved successfully. Shall never occur.");
 
 #ifdef CONFIG_DEBUG_COREDUMP
-	z_arm_coredump_fault_sp = POINTER_TO_UINT(esf);
```
Since there's already a structure to hold the ESF, wouldn't it make sense to add SP to it directly rather than using a global variable instead ?

arbrauns replied:
It would have to go in `extra_info` since it's not part of the hardware ESF, and then it would *only* be available with `CONFIG_EXTRA_EXCEPTION_INFO=y`, even though it's pretty basic and critical information.

ithinuel replied:
is there anything preventing from having 
```c
#if defined(CONFIG_DEBUG_COREDUMP)
	uint32_t fault_sp;
#endif
```

in `struct arch_esf` between the FPU stack frame and the extra esf info?

Note I don't know where the coredump expect SP to be pointing at so I can only assume that your change is correct.
But this made me realise that this global variable unnecessarily reserves a word of memory for no use beyond the few milli(micro?) seconds the fault handler goes from `z_arm_fault` down to `coredump` when all other data are passed between them through `arch_esf`.

arbrauns replied:
You could probably do something like that, but it would require further changes in pretty critical code that I'm not familiar with, so I don't really dare to touch this for a single word of memory usage.

ithinuel commented on include/zephyr/arch/arm/cortex_m/exception.h:
```diff
@@ -119,6 +119,15 @@ struct arch_esf {
+ * exception is the ESF base address plus this size.
+ */
+#if defined(CONFIG_EXTRA_EXCEPTION_INFO)
+#define Z_ARCH_HW_ESF_SIZE offsetof(arch_esf, extra_info)
```
This might need to be name Z_ARM_* rather than Z_ARCH as this is not really a zephyr Arch thing but a very arm support specific define.

arbrauns replied:
Renamed to `Z_ARM_MACHINE_ESF_SIZE`.

arbrauns commented:
Marking as draft because it needs more logic - the size of the Exception Stack Frame depends on whether or not floating-point instructions were used by the faulting thread (`CONTROL.FPCA`): https://developer.arm.com/documentation/107706/0100/Floating-point-and-MVE-support/Floating-point-context-handling-mechanisms?lang=en

When `CONTROL.FPCA` is clear, only `struct __basic_sf` is pushed; if it is set, both `struct __basic_sf` and `struct __fpu_sf` were pushed.

arbrauns committed:
arch: cortex_a_r: fix comment in swap code

This was introduced in ff1a5e78580da0ce2631acabab019ed92859b3ba, which
already put the ESF in r0 and the exception reason in r1.

Signed-off-by: Armin Brauns <armin.brauns@embedded-solutions.at>

arbrauns commented:
I don't understand the C++ errors (`invalid application of 'sizeof' to incomplete type 'z_arm_get_hw_esf_size(uint32_t)::__basic_sf'`).

mrkhldn commented:
I believe I'm trying to address the same issue here - https://github.com/zephyrproject-rtos/zephyr/pull/73189

arbrauns commented:
Hah, indeed, left a review there.

MaureenHelm commented:
@ithinuel please take a look

kartben commented:
@arbrauns this needs rebasing. Thanks!

arbrauns commented:
Fixed by #79622.
